### PR TITLE
Fix memory leak in checkpointer process

### PR DESCRIPTION
**Pull request by gfphoenix78**
The CurrentMemoryContext in CreateCheckPoint is long-live
memory context owned by the checkpointer process. The memory context
is reset only if the error occurs. So memory leak will lead to
huge memory leak in the OS.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
It seems like only the first 2 rdata are xloged. The remaining `prepared_transaction_agg_state` is ignored accidentally. This may be caused by the old code, but the items 2,3,4 are removed.
Am I wrong?

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -8919,16 +8919,15 @@ CreateCheckPoint(int flags)
 	rdata[1].data = (char *) dtxCheckPointInfo;
 	rdata[1].len = dtxCheckPointInfoSize;
 	rdata[1].buffer = InvalidBuffer;
-	rdata[1].next = NULL;
+	rdata[1].next = &(rdata[2]);
```
Comment (pivotal-ning-yu) [inline src/backend/access/transam/xlog.c]: This part of change is not for the memory leak, and it affects the logic in `XLogInsert_Internal()`, could we move it to a separate commit and explain it a bit?

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -8952,6 +8951,12 @@ CreateCheckPoint(int flags)
 
 	XLogFlush(recptr);
 
+	/* pfree memory blocks used by XLog to avoid memory leak in checkpointer. */
```
Comment (pivotal-ning-yu) [inline src/backend/access/transam/xlog.c]: The comment is boring, in fact I think it's unnecessary to explain why we want to free some memory, it's always to avoid memory leaks.

File: src/backend/access/transam/xlog.c [hunk 3]
```diff
@@ -8952,6 +8951,12 @@ CreateCheckPoint(int flags)
...
+	pfree(dtxCheckPointInfo);
+	pfree(p);
+	dtxCheckPointInfo = NULL;
+	p = NULL;
```
Comment (pivotal-ning-yu) [inline src/backend/access/transam/xlog.c]: Setting the pointers to NULL after freeing is good, however the real users of the memory are `rdata`, so instead of nullifying these two pointers it can be more useful to put a comment like `do not access rdata since now`.

File: src/backend/access/transam/xlog.c (hunk 2 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: It tells people the current memory context will not reset in the near future, don't rely on the memory context's reset.

File: src/backend/access/transam/xlog.c (hunk 1 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c]: Yes, this is not related to the memory leak issue.
I didn't see where `rdata[1].next` is set, but the fifth item is set and unused.

File: src/backend/access/transam/xlog.c [hunk 4]
```diff
@@ -8922,14 +8922,7 @@ CreateCheckPoint(int flags)
 	rdata[1].next = NULL;
 
 	prepared_transaction_agg_state *p = NULL;
-
 	getTwoPhasePreparedTransactionData(&p);
-	rdata[5].data = (char*)p;
```
Comment (asimrp) [inline src/backend/access/transam/xlog.c:8927]: With this line gone, we no longer are recording prepared transactions in checkpoint.  Wouldn't that break something?  E.g. if a checkpoint occurs after a transaction wrote its prepare record but has not yet received finish prepared from QD.  And the segment crashes.  Crash recovery will not replay the prepare record because it precedes most recent checkpoint record.  How can this transaction be finished?

Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c:8927]: Actually, 6X doesn't record the rdata[5]. You see, `rdata[0].next = &rdata[1], rdata[1].next = NULL`. I'm doubting here. I'm not clear how to write test case to verify it.

File: src/test/isolation2/sql/crash_recovery_dtm.sql [hunk 5]
```diff
@@ -138,7 +147,44 @@ $$ LANGUAGE plpgsql;
...
+1: CHECKPOINT;
+3: BEGIN;
+3: INSERT INTO QE_panic_test_table SELECT i, i+1 from generate_series(7, 12) i;
+3: CREATE TABLE QE_panic_test_table2(i int);
+3&: END;
+
+1: SELECT gp_wait_until_triggered_fault('finish_prepared_start_of_function', 2, dbid)
+   from gp_segment_configuration where role='p' and content=0;
+1: SELECT pg_ctl(datadir, 'restart') from gp_segment_configuration where role = 'p' and content = 0;
+4: SELECT pg_sleep(2);
```
Comment (xiong-gang) [inline src/test/isolation2/sql/crash_recovery_dtm.sql:170]: 2 seconds is not promising if the system is very slow, it's better to check the segment has finished the recovery.

Comment (gfphoenix78) [reply src/test/isolation2/sql/crash_recovery_dtm.sql:170]: Yes, I agree.

File: src/backend/access/transam/xlog.c (hunk 4 above)
Comment (gfphoenix78) [reply src/backend/access/transam/xlog.c:8927]: Fixed

File: src/test/isolation2/sql/crash_recovery_dtm.sql (hunk 5 above)
Comment (gfphoenix78) [reply src/test/isolation2/sql/crash_recovery_dtm.sql:170]: There is no sleep/wait in some places. It may be the `pg_ctl`'s responsibility to take care. I'll remove the sleep clause. If it's a problem, we should fix it in the UDF `pg_ctl`.
How do you think about it? @xiong-gang
